01-architecture.md

```diff
@@ -342,6 +342,27 @@ This is especially important in a sharded blockchain design, where actor code an
 
 Gas accounting will be performed at the bytecode level, leveraging the metering facilites provided by the WASM runtimes under consideration (Wasmer, Wasmtime).
 
+## Actor code updates
+
+The FVM aims to support updating actor code. This choice diverges from the "code immutability" paradigm that Ethereum made popular. That paradigm is empowering on one hand (the user has the guarantee to always be dealing with the same code), but limiting on the other (software is never static).
+
+The reality is that software evolves continuously driven by feature development, improvements, bug fixing, security patching, and software maintenance (adaptive, corrective, preventive, perfective). Smart contracts _are_ software, and they are not exempted from the need of evolution.
+
+To work around the EVM limitations, Ethereum developers have widely adopted the ["upgradable proxy"](https://docs.openzeppelin.com/upgrades-plugins/1.x/proxies), but this comes at the cost of extra complexity, boilerplate, and yet another point of failure. Most Dapps adopt this pattern.
+
+In Filecoin, we rely on content-addressed code as a foundational building block to enable code upgradability. Actors have an address, but code is not deployed at that address (like in the EVM); instead it is linked to a CID in the state tree.
+
+This decoupling makes it feasible to atomically update actor code. Doing so would imply:
+
+1. Loading new actor code and obtaining a code CID; see [Actor deployment](#actor-deployment)
+2. Calling an `UpdateCode` syscall from within the actor, with the new code CID.
+
+All future messages directed to this actor will use the new code.
+
+Authentication and security behind code updates are entirely in actor space; therefore, actors are free to implement the checks, guards and logic that ultimately lead to the invocation of the `UpdateCode` syscall.
```

Comment:
I'd also note that it's possible to statically determine that an actor _can't_ call this function.